### PR TITLE
Add Utf8JsonWriter.WriteRawValue(System.Buffers.ReadOnlySequence).

### LINKED ISSUE #68223: [API Proposal]: Add ReadOnlySequence overload for WriteRawValue
### Background and motivation

When using the `Utf8JsonWriter.WriteRawValue` API, you can't pass multiple segments in to write a single value, you have to allocate a contiguous array and do a single call to `WriteRawValue`. In our use case we already have a `ReadOnlySequence<byte>` and want to write it as a raw json value.

But because the API only accepts string or span, we need to write the following:
```c#
if (result.RawSerializedData.IsSingleSegment)
{
    writer.WriteRawValue(result.RawSerializedData.First.Span, skipInputValidation: true);
}
else
{
    writer.WriteRawValue(result.RawSerializedData.ToArray(), skipInputValidation: true);
}
```

### API Proposal

```diff
namespace System.Text.Json
{
    public sealed partial class Utf8JsonWriter
    {
+        public void WriteRawValue(ReadOnlySequence<byte> json, bool skipInputValidation = false);
+        public void WriteRawValue(ReadOnlySequence<char> json, bool skipInputValidation = false);
    }
}
```

### API Usage

```C#
void MyFunc(Utf8JsonWriter writer, ReadOnlySequence<byte> sequence)
{
    writer.WriteRawValue(sequence);
}
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by lateapexearlyspeed**
Fix #68223

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 1]
```diff
@@ -194,9 +263,37 @@ private void WriteRawValueCore(ReadOnlySpan<byte> utf8Json, bool skipInputValida
             }
 
             utf8Json.CopyTo(output.Slice(BytesPending));
-            BytesPending += len;
+            BytesPending += (int)len;
 
             SetFlagToAddListSeparatorBeforeNextItem();
         }
+
+        private void GrowIfNeed(int len)
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:271]: nit: `GrowIfNeeded`

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs [hunk 2]
```diff
@@ -450,22 +461,27 @@ private enum OverloadParamType
...
+                    writer.WriteRawValue(payload, skipInputValidation);
                     break;
                 case OverloadParamType.ByteArray:
                     byte[] payloadAsBytes = Encoding.UTF8.GetBytes(payload);
-                    writer.WriteRawValue(payloadAsBytes);
+                    writer.WriteRawValue(payloadAsBytes, skipInputValidation);
+                    break;
+                case OverloadParamType.ROSeqByte:
+                    ReadOnlySequence<byte> payloadAsSequence = new(Encoding.UTF8.GetBytes(payload));
+                    writer.WriteRawValue(payloadAsSequence, skipInputValidation);
```
Comment (krwq) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs:484]: small suggestion: you might wanna add Debug.Fail in the default case of switch statement to future proof this code

**Comment by Wanath** [comment]
ترخيص MIT

حقوق الطبع والنشر (c) [year] [fullname]

يُمنح الإذن بموجب هذا مجانًا لأي شخص يحصل على نسخة
من هذا البرنامج وملفات التوثيق المرتبطة به ("البرنامج") ، للتعامل
في البرنامج دون قيود ، بما في ذلك على سبيل المثال لا الحصر الحقوق
لاستخدام و / أو نسخ أو تعديل أو دمج أو نشر أو توزيع أو ترخيص فرعي و / أو بيع
نسخ من البرنامج ، والسماح للأشخاص الذين يكون البرنامج لهم
مفروشة للقيام بذلك ، مع مراعاة الشروط التالية:

يجب تضمين إشعار حقوق النشر أعلاه وإشعار الإذن هذا في الكل
نسخ أو أجزاء كبيرة من البرنامج.

يتم توفير البرنامج "كما هو" ، بدون أي ضمان من أي نوع ، صريحًا أو
ضمنيًا ، بما في ذلك على سبيل المثال لا الحصر ضمانات القابلية للتسويق ،
الملاءمة لغرض معين وعدم الانتهاك. لا يجوز في أي حال من الأحوال أن
المؤلفون أو حاملو حقوق الطبع والنشر مسؤولون عن أي مطالبة أو أضرار أو غير ذلك
المسؤولية ، سواء كانت ناتجة عن عقد أو ضرر أو غير ذلك ، تنشأ عن ،
خارج أو متصل بالبرنامج أو الاستخدام أو المعاملات الأخرى في
البرمجيات.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 3]
```diff
@@ -194,9 +263,37 @@ private void WriteRawValueCore(ReadOnlySpan<byte> utf8Json, bool skipInputValida
...
+        {
+            // TODO (https://github.com/dotnet/runtime/issues/29293):
```
[the remark this answers is not in the document]
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Removed.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 1 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Fixed.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 4]
```diff
@@ -113,6 +113,42 @@ public void WriteRawValue(ReadOnlySpan<byte> utf8Json, bool skipInputValidation
...
+        /// The <see cref="JsonWriterOptions.Indented"/> and <see cref="JsonWriterOptions.Encoder"/> values for the writer instance are not applied when using this method.
+        /// </remarks>
+        public void WriteRawValue(ReadOnlySequence<byte> utf8Json, bool skipInputValidation = false)
+        {
+            if (!_options.SkipValidation)
+            {
+                ValidateWritingValue();
+            }
+
+            if (utf8Json.Length == int.MaxValue)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:144]: I know this check has been taken from the `ReadOnlySpan` overload, but I struggle to understand what it means. Why only check for `int.MaxValue` lengths specifically?

For the case of `ReadOnlySpan`, it will almost always be false (as managed buffers are bounded by `Array.MaxLength` which in turn is less than `int.MaxValue`). For the case of `ReadOnlySequence` the `Length` property is `long` so any value `> int.MaxValue` will happily pass this check.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:144]: cc @layomia

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 5]
```diff
@@ -194,9 +263,35 @@ private void WriteRawValueCore(ReadOnlySpan<byte> utf8Json, bool skipInputValida
...
+        {
+            int maxRequired = len + 1; // Optionally, 1 list separator. We've guarded against integer overflow earlier in the call stack.
+
+            if (_memory.Length - BytesPending < maxRequired)
+            {
+                Grow(maxRequired);
+            }
+        }
+
+        private static JsonTokenType ValidateJson(in ReadOnlySpan<byte> utf8Json)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:281]: The `ValidateJson` methods don't seem particularly justified given that they're only being called once each.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 6]
```diff
@@ -194,9 +263,35 @@ private void WriteRawValueCore(ReadOnlySpan<byte> utf8Json, bool skipInputValida
             }
 
             utf8Json.CopyTo(output.Slice(BytesPending));
-            BytesPending += len;
+            BytesPending += (int)len;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:266]: Arguably, this should be a `checked` cast and should have happened much earlier in the validation stage.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 7]
```diff
@@ -194,9 +263,35 @@ private void WriteRawValueCore(ReadOnlySpan<byte> utf8Json, bool skipInputValida
...
 
             SetFlagToAddListSeparatorBeforeNextItem();
         }
+
+        private void GrowIfNeeded(int len)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:271]: Why is this added method necessary?

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 4 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Fixed.
For `ReadOnlySequence `overload side, it is my oversight when I thought they are similar codes at first sight, sorry about that ...
For `ReadOnlySpan `overload side, I guess this int length cannot be int.MaxValue because [subsequent code](https://github.com/dotnet/runtime/blob/1281a4a3393e9293aa4b379d619a692d6dcb62a4/src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs#L182) may add `','`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 5 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Removed.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 8]
```diff
@@ -171,21 +207,54 @@ private void WriteRawValueCore(ReadOnlySpan<byte> utf8Json, bool skipInputValida
...
+            if (len == 0)
+            {
+                ThrowHelper.ThrowArgumentException(SR.ExpectedJsonTokens);
             }
 
+            // In the UTF-16-based entry point methods above, we validate that the payload length <= int.MaxValue /3.
+            // The result of this division will be rounded down, so even if every input character needs to be transcoded
+            // (with expansion factor of 3), the resulting payload would be less than int.MaxValue,
+            // as (int.MaxValue/3) * 3 is less than int.MaxValue.
+            Debug.Assert(len < int.MaxValue);
```
[the remark this answers is not in the document]
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Validation added.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 6 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Added check in earlier validation stage.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 7 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Removed.
Added previously because I cannot figure out better way to reduce duplicated code between `ReadOnlySpan `and `ReadOnlySequence `overloads, so just make them shorter in main methods.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs [hunk 9]
```diff
@@ -244,7 +254,7 @@ public static void WriteRawNullOrEmptyTokenInvalid()
             Assert.Throws<ArgumentNullException>(() => writer.WriteRawValue(json: default(string)));
             Assert.Throws<ArgumentException>(() => writer.WriteRawValue(json: ""));
             Assert.Throws<ArgumentException>(() => writer.WriteRawValue(json: default(ReadOnlySpan<char>)));
-            Assert.Throws<ArgumentException>(() => writer.WriteRawValue(utf8Json: default));
+            Assert.Throws<ArgumentException>(() => writer.WriteRawValue(utf8Json: default(ReadOnlySpan<byte>)));
```
[the remark this answers is not in the document]
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs:257]: Added.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 10]
```diff
@@ -173,12 +249,10 @@ private void WriteRawValueCore(ReadOnlySpan<byte> utf8Json, bool skipInputValida
             {
                 // Utilize reader validation.
                 Utf8JsonReader reader = new(utf8Json);
-                while (reader.Read());
+                while (reader.Read()) ;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:252]:
```suggestion
                while (reader.Read());
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 11]
```diff
@@ -113,6 +113,82 @@ public void WriteRawValue(ReadOnlySpan<byte> utf8Json, bool skipInputValidation
...
+            else
+            {
+                // Utilize reader validation.
+                Utf8JsonReader reader = new(utf8Json);
+                while (reader.Read()) ;
+                _tokenType = reader.TokenType;
+            }
+
+            Debug.Assert(len < int.MaxValue);
+            int maxRequired = (int)len + 1; // Optionally, 1 list separator. We've guarded against integer overflow earlier in the call stack.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:172]: `len` is being cast a couple of times in this method body, so it's probably preferable if you did it once immediately after the `if (len >= int.MaxValue)` check and use the result everywhere else.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 12]
```diff
@@ -113,6 +113,82 @@ public void WriteRawValue(ReadOnlySpan<byte> utf8Json, bool skipInputValidation
...
+
+            long len = utf8Json.Length;
+
+            if (len == 0)
+            {
+                ThrowHelper.ThrowArgumentException(SR.ExpectedJsonTokens);
+            }
+            if (len >= int.MaxValue)
+            {
+                ThrowHelper.ThrowArgumentException_ValueTooLarge(len);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:152]: Is this expected behavior for ReadOnlySequence inputs larger than that? Do we have other methods that accept ROS? What is prior art in this space?

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 13]
```diff
@@ -113,6 +113,82 @@ public void WriteRawValue(ReadOnlySpan<byte> utf8Json, bool skipInputValidation
...
+                // not affect structural validation since a string token is equivalent to a complete object, array,
+                // or other complete JSON tokens when considering structural validation on subsequent writer calls.
+                // If the payload is not valid, then we make no guarantees about the structural validation of the final payload.
+                _tokenType = JsonTokenType.String;
+            }
+            else
+            {
+                // Utilize reader validation.
+                Utf8JsonReader reader = new(utf8Json);
+                while (reader.Read()) ;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:167]: Nit:
```suggestion
                while (reader.Read());
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 14]
```diff
@@ -173,12 +249,10 @@ private void WriteRawValueCore(ReadOnlySpan<byte> utf8Json, bool skipInputValida
             {
                 // Utilize reader validation.
                 Utf8JsonReader reader = new(utf8Json);
-                while (reader.Read());
+                while (reader.Read()) ;
                 _tokenType = reader.TokenType;
             }
 
-            // TODO (https://github.com/dotnet/runtime/issues/29293):
-            // investigate writing this in chunks, rather than requesting one potentially long, contiguous buffer.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:181]: Curious why this got removed? Was this addressed here?

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Issue was closed automatically because of no activity. Unless the concern has been addressed with this PR, we should leave the comment (even if it might be pointing to a closed issue). Closed issues can be reopened, context from deleted comments is hard to recover.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 14 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Added back per [comment](https://github.com/dotnet/runtime/pull/76444#discussion_r987857215)

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 13 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Fixed.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 11 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Fixed.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 4 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Ok, just confirm, we will see @layomia 's reply about that check.

Comment (jeffhandley) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: This is the only remaining question on this PR--whether the check for `len == int.MaxValue` is a valuable defense against an overflow exception.

@lateapexearlyspeed To determine if the guard is useful or not, could you try adding a unit test that would try to hit that boundary case? Let's see what happens with and without the guard. A similar test could be added for the `ReadOnlySpan` overload too.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: @lateapexearlyspeed I would make the following recommendations:

1. Completely remove the `len == int.MaxValue` check (and update API docs above) from the existing span overload.
2. Update the check in the new ReadOnlySequence overload to say `len <= int.MaxValue` (so that the length can be safely cast to an `int32`).
3. Update the assertion in [the core methods](https://github.com/dotnet/runtime/blob/a715f30d106f69e6243c46f82a3cacd12568936d/src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs#L237-L241) to say `len <= int.MaxValue`.
4. Add unit tests covering these boundary conditions.

Thanks!

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: > Add unit tests covering these boundary conditions

I believe these are covered here but it's still worth some thought https://github.com/dotnet/runtime/blob/af8694704e807d0147a3033c3e250be2966951fe/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs#L478

Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: For new ROSeq, there was also [test case](https://github.com/dotnet/runtime/blob/a715f30d106f69e6243c46f82a3cacd12568936d/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs#L532) added to cover boundary conditions.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs [hunk 15]
```diff
@@ -141,15 +141,15 @@ public void WriteRawValue(ReadOnlySequence<byte> utf8Json, bool skipInputValidat
                 ValidateWritingValue();
             }
 
-            long len = utf8Json.Length;
+            int len = checked((int)utf8Json.Length);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:144]: The cast should be moved after the check so that large inputs always always surface as the same exception (`ArgumentException` rather than `OverflowException`).

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs [hunk 16]
```diff
@@ -536,7 +536,15 @@ public void WriteRawUtf8LengthGreaterThanOrEqualToIntMax(long len)
                 using MemoryStream ms = new();
                 using Utf8JsonWriter writer = new(ms);
                 ReadOnlySequence<byte> readonlySeq = CreateLargeReadOnlySequence(len);
-                Assert.Throws<ArgumentException>(() => writer.WriteRawValue(readonlySeq));
+
+                if (len == int.MaxValue)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs:540]: Per feedback above, this should always throw `ArgumentException`

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: > It wasn't.
> I recommend we leave as-is. We require the the input length is than `int.MaxValue` so we can add a comma if required. The comment talks about validation in the Utf16 based entry points (the info about integer division) but should also state that the ROSpan/Sequence entry points need to adhere to this to.

I'm not sure I follow. Even you are able to add one comma at the end, that isn't necessary to produce a complete JSON document. 

My concern is that in 99.9% of use cases, this check isn't particularly meaningful since both the source and `Utf8JsonWriter` destination buffers are bounded by `Array.MaxLength` which is substantially less than `int.MaxValue`. You could, in theory, construct an unmanaged span of length `int.MaxValue`, but if you do that then your code probably has bigger problems than not being able to append one character at the end.

In any case, I probably agree we should not try to change this in the context of this PR.

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: > Utf8JsonWriter destination buffers are bounded by Array.MaxLength which is substantially less than int.MaxValue.

Yeah that's right - context I didn't keep in mind when thinking about this.

> we should not try to change this in the context of this PR.

👍 yeah fine to address in a follow up.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 15 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:144]: This single commit change was just because of [comment](https://github.com/dotnet/runtime/pull/76444#discussion_r987853288), checked cast seems prior art for ROSeq.Length in this space.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs (hunk 16 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.WriteRaw.cs:540]: See [comment](https://github.com/dotnet/runtime/pull/76444/files/134cbd9b3f366b1fd9d219163372c9c99c7a3f93..a715f30d106f69e6243c46f82a3cacd12568936d#r1056365655).

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs (hunk 15 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs:144]: In any case, we are throwing two different exceptions for what is essentially the same failure mode. I would recommend making it just one (and preferably ArgumentException since we can control the error message, unlike checked casts).

Comment (lateapexearlyspeed) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs]: Fixed.

**Review by eiriktsarpalis — APPROVED** [review]
LGTM, thanks! I've rebased your changes on top of the latest main, will merge once CI completes.

**Comment by eiriktsarpalis** [comment]
Thank you for your contribution @lateapexearlyspeed!
